<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>co-expression module perturbation analysis • compact</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="co-expression module perturbation analysis">
<meta name="description" content="Tutorial for performing module perturbation analysis (MPA) with compact.
">
<meta property="og:description" content="Tutorial for performing module perturbation analysis (MPA) with compact.
">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">compact</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/basic_tutorial.html">co-expression module perturbation analysis</a></li>
    <li><a class="dropdown-item" href="../articles/TF_tutorial.html">transcription factor perturbation analysis</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>co-expression module perturbation analysis</h1>
            
      

      <div class="d-none name"><code>basic_tutorial.Rmd</code></div>
    </div>

    
    
<p>Compiled: 28-10-2024</p>
<p>Source: <code>vignettes/basic_tutorial.Rmd</code></p>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This tutorial covers the basics of using
<strong><code>compact</code></strong> to perform module perturbation
analysis on co-expression network/module on single-cell data.</p>
<p>This tutorial covers the basics of using
<strong><code>compact</code></strong> to perform in-silico pertubation
experiments in single-cell transcriptomics data. For this tutorial we
will use an integrated single-nucleus RNA-seq dataset of microglia from
three different studies of Alzheimer’s disease, as we described
previously in the <a href="https://www.sciencedirect.com/science/article/pii/S2667237523001273?via%3Dihub" class="external-link">hdWGCNA
paper</a>.</p>
<p>The in-silico perturbations that we demonstrate in this tutorial are
based on gene co-expression modules, which are unbiased clusters of
genes that are highly co-expressed in a gene-gene co-expression network.
In order to run these pertubations, we first need to perform
co-expression network analysis and group genes into modules by running
<a href="https://smorabit.github.io/hdWGCNA/articles/basic_tutorial.html" class="external-link">hdWGCNA</a>.
This has already been run on the tutorial microglia dataset, where we
have identified four gene modules.</p>
</div>
<div class="section level2">
<h2 id="load-the-dataset">Load the dataset<a class="anchor" aria-label="anchor" href="#load-the-dataset"></a>
</h2>
<p>Download the tutorial dataset from these Google Drive links:</p>
<ul>
<li><p><a href="https://drive.google.com/file/d/15-0URxGUgE2eCxx15zkoVwkkmLhZTFKH/view?usp=drive_link" class="external-link">Seurat
object</a></p></li>
<li><p><a href="https://drive.google.com/file/d/1BcnUx3eh4zHwk_JfBo8zpB70ewp0u1tX/view?usp=drive_link" class="external-link">co-expression
adjacency matrix (.rda)</a></p></li>
</ul>
<p>First we load the required R libraries and the tutorial dataset. When
loading the tutorial dataset, you need to update the filepath to the
co-expression adjacency matrix (TOM) within the Seurat object (shown
below).</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/" class="external-link">cowplot</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">velocyto.R</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://smorabit.github.io/hdWGCNA/" class="external-link">hdWGCNA</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://smorabit.github.io/COMPACT/" class="external-link">compact</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/get_theme.html" class="external-link">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/theme_cowplot.html" class="external-link">theme_cowplot</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># set random seed for reproducibility</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">12345</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># load the processed alzheimer's microglia dataset</span></span>
<span><span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span>file<span class="op">=</span><span class="st">'AD_MG_hdWGCNA_COMPACT_tutorial.rds'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># update the path to the co-expression network adjacency matrix</span></span>
<span><span class="va">net</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://smorabit.github.io/hdWGCNA/reference/GetNetworkData.html" class="external-link">GetNetworkData</a></span><span class="op">(</span><span class="va">seurat_obj</span><span class="op">)</span></span>
<span><span class="va">net</span><span class="op">$</span><span class="va">TOMFiles</span> <span class="op">&lt;-</span> <span class="st">'AD_MG_TOM.rda'</span> <span class="co"># it is better to use the absolute path</span></span>
<span><span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://smorabit.github.io/hdWGCNA/reference/SetNetworkData.html" class="external-link">SetNetworkData</a></span><span class="op">(</span><span class="va">seurat_obj</span>, <span class="va">net</span><span class="op">)</span></span></code></pre></div>
<p>If you have not already done so, we need to construct a
nearest-neighbors graph in the Seurat object. This is a key step in the
typical single-cell clustering pipeline.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a>seurat_obj <span class="ot">&lt;-</span> <span class="fu">FindNeighbors</span>(</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a>    seurat_obj,</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a>    <span class="at">reduction=</span><span class="st">'harmony'</span>,nearest neighbors?</span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a>    <span class="at">assay =</span> <span class="st">'RNA'</span>,</span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a>    <span class="at">annoy.metric =</span> <span class="st">'cosine'</span></span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a>)</span></code></pre></div>
<p>Next we visualize the clusters in this dataset as a UMAP plot.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get the UMAP embedding from the seurat object</span></span>
<span><span class="va">emb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/ObjectAccess.html" class="external-link">Reductions</a></span><span class="op">(</span><span class="va">seurat_obj</span>, <span class="st">'umap'</span><span class="op">)</span><span class="op">@</span><span class="va">cell.embeddings</span></span>
<span></span>
<span><span class="co"># make a dataframe to plot with ggplot</span></span>
<span><span class="va">plot_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">emb</span><span class="op">)</span></span>
<span><span class="va">plot_df</span><span class="op">$</span><span class="va">functional_anno</span> <span class="op">&lt;-</span> <span class="va">seurat_obj</span><span class="op">$</span><span class="va">functional_anno</span></span>
<span></span>
<span><span class="co"># plot with ggplot</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="va">plot_df</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">UMAP_1</span>, y<span class="op">=</span><span class="va">UMAP_2</span>, color<span class="op">=</span><span class="va">functional_anno</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_color_viridis_d</a></span><span class="op">(</span>option <span class="op">=</span> <span class="st">"plasma"</span><span class="op">)</span> <span class="op">+</span>  </span>
<span>  <span class="fu">rasterise</span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">2</span>, alpha<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span>, dpi <span class="op">=</span> <span class="fl">300</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"microglia functional_anno"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">hdWGCNA</span><span class="fu">::</span><span class="fu"><a href="https://smorabit.github.io/hdWGCNA/reference/umap_theme.html" class="external-link">umap_theme</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p</span></span></code></pre></div>
<p><img src="figures/basic_tutorial/UMAP_microglia_functional_anno_MPA.png" width="1000" height="2000"></p>
</div>
<div class="section level2">
<h2 id="perform-in-silico-module-perturbation">Perform in-silico module perturbation<a class="anchor" aria-label="anchor" href="#perform-in-silico-module-perturbation"></a>
</h2>
<p>In this example, we will perform an in-silico knock-down for one of
our microglia co-expression modules. First, let’s look at the expression
level of these modules (module eigengene).</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># add module expression levels (MEs) to the seurat metadata</span></span>
<span><span class="va">MEs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://smorabit.github.io/hdWGCNA/reference/GetMEs.html" class="external-link">GetMEs</a></span><span class="op">(</span><span class="va">seurat_obj</span><span class="op">)</span></span>
<span><span class="va">meta</span> <span class="op">&lt;-</span> <span class="va">seurat_obj</span><span class="op">@</span><span class="va">meta.data</span></span>
<span><span class="va">seurat_obj</span><span class="op">@</span><span class="va">meta.data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">meta</span>, <span class="va">MEs</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># dotplot</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DotPlot.html" class="external-link">DotPlot</a></span><span class="op">(</span></span>
<span>    <span class="va">seurat_obj</span>, features<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'MG-M4'</span>, <span class="st">'MG-M1'</span>, <span class="st">'MG-M3'</span>, <span class="st">'MG-M2'</span><span class="op">)</span>, </span>
<span>    group.by<span class="op">=</span><span class="st">'functional_anno'</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_color_gradient2</a></span><span class="op">(</span>high<span class="op">=</span><span class="st">'black'</span>, mid<span class="op">=</span><span class="st">'grey75'</span>, low<span class="op">=</span><span class="st">'grey99'</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>        axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>        axis.ticks.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">'Trajectory →'</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">''</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoLegend</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p</span></span></code></pre></div>
<p><img src="figures/basic_tutorial/MG_module_dotplot.png" width="1000" height="1000"></p>
<p>The expression of microglia module M4 is concentrated at the end of
the trajectory, within the activated microglia population. Next we will
perform a knock-down experiment of this module, and we will predict how
this perturbation will alter cellular states.</p>
<div class="section level3">
<h3 id="apply-a-knock-down-using-moduleperturbation">Apply a knock-down using <code>ModulePerturbation</code><a class="anchor" aria-label="anchor" href="#apply-a-knock-down-using-moduleperturbation"></a>
</h3>
<p>Here we use the function <code>ModulePerturbation</code> to apply an
in-silico knock-down of module MG-M4. This function consists of three
main steps, which are all conveniently included in this function.</p>
<ul>
<li><p>Apply a primary perturbation to the hub genes of the
co-expression module. For each hub gene, we perform zero-inflated count
data regression to model the observed gene expression distribution. We
simulate a perturbation by sampling from this distribution, then adding
or subtracting the sampled counts to the observed counts for a knock-in
or a knock-down respectively.</p></li>
<li><p>Apply a secondary perturbation to all other genes in the
co-expression module. The primary perturbation is propagated to other
genes by exploiting the co-expression network structure. Genes with
stronger connections in the network will exhibit a stronger
perturbation.</p></li>
<li><p>Calculate cell-cell transition probabilities based on the
observed gene expression matrix and the perturbed gene expression
matrix.</p></li>
</ul>
<p>For this analysis we will perform the primary perturbation on the top
10 hub genes of module MG-M4. Let’s check which genes we will be
perturbing.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hub_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://smorabit.github.io/hdWGCNA/reference/GetHubGenes.html" class="external-link">GetHubGenes</a></span><span class="op">(</span><span class="va">seurat_obj</span>, n_hubs<span class="op">=</span><span class="fl">10</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">module</span> <span class="op">==</span> <span class="st">'MG-M4'</span><span class="op">)</span></span>
<span><span class="va">hub_genes</span><span class="op">$</span><span class="va">gene_name</span></span></code></pre></div>
<pre><code>[1] "PTPN2"   "SLC11A1" "ACSL1"   "SPATA6"  "TBC1D14" "TMEM163" "ERC2"   
[8] "DPYD"    "CD163"   "TLR2"   </code></pre>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># provide a name for the perturbation, which will be stored in the Seurat object</span></span>
<span><span class="va">perturbation_name</span> <span class="op">&lt;-</span> <span class="st">'M4_down'</span></span>
<span></span>
<span><span class="co"># run a knock-down of the top 10 hub genes from MG-M4</span></span>
<span><span class="co"># should take 2-3 min to run</span></span>
<span><span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ModulePerturbation.html">ModulePerturbation</a></span><span class="op">(</span></span>
<span>    <span class="va">seurat_obj</span>,</span>
<span>    mod <span class="op">=</span> <span class="st">"MG-M4"</span>,</span>
<span>    perturb_dir <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>, <span class="co"># negative indicates knock-down, positive indicates knock-in</span></span>
<span>    perturbation_name <span class="op">=</span> <span class="va">perturbation_name</span>,</span>
<span>    graph <span class="op">=</span> <span class="st">'RNA_nn'</span>, </span>
<span>    n_hubs <span class="op">=</span> <span class="fl">10</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The main ouputs of this function are a “perturbed” gene expression
matrix, and a cell-cell transition matrix, which we can access from the
Seurat object using the following code.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get the perturbed expression matrix</span></span>
<span><span class="va">X_per</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/AssayData.html" class="external-link">GetAssayData</a></span><span class="op">(</span></span>
<span>  <span class="va">seurat_obj</span>, assay<span class="op">=</span><span class="va">perturbation_name</span>,</span>
<span>  layer<span class="op">=</span><span class="st">'counts'</span> </span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># get the cell-cell transition matrix</span></span>
<span><span class="va">tp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/ObjectAccess.html" class="external-link">Graphs</a></span><span class="op">(</span><span class="va">seurat_obj</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">perturbation_name</span>, <span class="st">'_tp'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Next, let’s compare the observed and perturbed gene expression
distributions for one of the MG-M4 hub genes, TMEM163. We expect that in
the pertubed matrix, the expression of TMEM163 will be lower than in the
observed matrix.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Generate 10 colors from the plasma palette</span></span>
<span><span class="va">plasma_colors</span> <span class="op">&lt;-</span> <span class="fu">viridis</span><span class="fu">::</span><span class="fu"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">viridis</a></span><span class="op">(</span><span class="fl">10</span>, option <span class="op">=</span> <span class="st">"plasma"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># group variable for vln plots</span></span>
<span><span class="va">group.by</span> <span class="op">&lt;-</span> <span class="st">'functional_anno'</span></span>
<span></span>
<span><span class="co"># one of the hub genes to be perturbed</span></span>
<span><span class="va">cur_gene</span> <span class="op">&lt;-</span> <span class="st">'TMEM163'</span></span>
<span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html" class="external-link">VlnPlot</a></span><span class="op">(</span><span class="va">seurat_obj</span>, features <span class="op">=</span> <span class="va">cur_gene</span>, pt.size<span class="op">=</span><span class="fl">0.1</span>, group.by <span class="op">=</span> <span class="va">group.by</span>, assay<span class="op">=</span><span class="st">'RNA'</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">plasma_colors</span><span class="op">)</span> <span class="op">+</span>  <span class="co"># Use the extracted colors here</span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoLegend</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">''</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">''</span><span class="op">)</span> <span class="op">+</span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">ylim</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>    axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    plot.margin<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">margin</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">0</span>,<span class="fl">0</span>,<span class="fl">0</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html" class="external-link">VlnPlot</a></span><span class="op">(</span><span class="va">seurat_obj</span>, features <span class="op">=</span> <span class="va">cur_gene</span>, pt.size<span class="op">=</span><span class="fl">0.1</span>, group.by <span class="op">=</span> <span class="va">group.by</span>, assay<span class="op">=</span><span class="va">perturbation_name</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">plasma_colors</span><span class="op">)</span> <span class="op">+</span>  <span class="co"># Use the extracted colors here</span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoLegend</a></span><span class="op">(</span><span class="op">)</span>  <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">''</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">ylim</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">5</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>    plot.margin<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">margin</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">0</span>,<span class="fl">0</span>,<span class="fl">0</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">''</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">'Functional Annotation / Pseudotime --&gt;'</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">p1</span> <span class="op">/</span> <span class="va">p2</span></span></code></pre></div>
<p><img src="figures/basic_tutorial/vln_compare_M4_down_TMEM163.png" width="2000" height="1500"></p>
</div>
<div class="section level3">
<h3 id="visualize-predicted-cell-transitions-after-perturbation">Visualize predicted cell transitions after perturbation<a class="anchor" aria-label="anchor" href="#visualize-predicted-cell-transitions-after-perturbation"></a>
</h3>
<p>Here we will visualize the predicted cell-cell transitions after the
in-silico knock-down of MG-M4. Similar to other methods like <a href="https://github.com/velocyto-team/velocyto.R" class="external-link">RNA Velocity</a> or
<a href="https://github.com/morris-lab/CellOracle" class="external-link">CellOracle</a>, we
visualize the inferred transitions as a vector field overlaid on a
dimensionality reduction. We include the function
<code>PlotTransitionVectors</code> to generate this plot. This function
leverages some code from the R package <code>velocyto.R</code>, however
in a future update we would like to remove this dependency since this
package can be difficult to install on some machines.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PlotTransitionVectors.html">PlotTransitionVectors</a></span><span class="op">(</span></span>
<span>    <span class="va">seurat_obj</span>,</span>
<span>    perturbation_name <span class="op">=</span> <span class="va">perturbation_name</span>,</span>
<span>    reduction <span class="op">=</span> <span class="st">'umap'</span>,</span>
<span>    color.by<span class="op">=</span><span class="st">'pseudotime'</span>,</span>
<span>    grid_resolution <span class="op">=</span> <span class="fl">25</span> </span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="va">p</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoLegend</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_color_gradientn</a></span><span class="op">(</span>colors<span class="op">=</span><span class="fu">viridis</span><span class="fu">::</span><span class="fu"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">plasma</a></span><span class="op">(</span><span class="fl">256</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_equal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="va">perturbation_name</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p</span></span></code></pre></div>
<p><img src="figures/basic_tutorial/vectorfield_M4_down_velocyto.png" width="2000" height="1500"></p>
<p>To generate the vector field, the cell-cell transition probabilities
computed for this perturbation are used to infer the cells with similar
expression profiles, and vectors are shown between these cells. In order
to avoid overcrowding the plot, these vectors are averaged in a grid
rather than showing a vector for each cell.</p>
</div>
<div class="section level3">
<h3 id="compare-cell-state-similarity-before-and-after-perturbation">Compare cell state similarity before and after perturbation<a class="anchor" aria-label="anchor" href="#compare-cell-state-similarity-before-and-after-perturbation"></a>
</h3>
<p>To further analyze the effect of the perturbation, we compute the
E-distance between different functional annotations within the original
and perturbed Seurat objects. The E-distance helps assess the similarity
between different functional cell states present in the dataset.</p>
<p>First, we calculate energy distance between clusters in the observed
gene expression matrix using the function
<code>ComputeDistance</code>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Calculate similarity distance for original scRNA-seq data, method selected "edist" - energy distance</span></span>
<span><span class="va">df_edist_original</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ComputeDistance.html">ComputeDistance</a></span><span class="op">(</span></span>
<span>  <span class="va">seurat_obj</span>, </span>
<span>  groupby <span class="op">=</span> <span class="st">'functional_anno'</span>, </span>
<span>  reduction <span class="op">=</span> <span class="st">'pca'</span>, </span>
<span>  method <span class="op">=</span> <span class="st">"edist"</span>, </span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>  </span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># # Define your custom order</span></span>
<span><span class="va">custom_order</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="va">seurat_obj</span><span class="op">$</span><span class="va">functional_anno</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Rearrange rows and columns of df_edist_observed based on the custom order</span></span>
<span><span class="va">df_edist_original_reordered</span> <span class="op">&lt;-</span> <span class="va">df_edist_original</span><span class="op">[</span><span class="va">custom_order</span>, <span class="va">custom_order</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># View the reordered dataframe</span></span>
<span><span class="va">df_edist_original_reordered</span></span></code></pre></div>
<details><summary>
Output
</summary><pre><code><span><span class="co">#               Homeostasis Surveillance1 Surveillance2 Surveillance3</span></span>
<span><span class="co"># Homeostasis      0.000000     10.140967     17.220091    12.1832043</span></span>
<span><span class="co"># Surveillance1   10.140967      0.000000      2.866413     3.1124290</span></span>
<span><span class="co"># Surveillance2   17.220091      2.866413      0.000000     1.2850047</span></span>
<span><span class="co"># Surveillance3   12.183204      3.112429      1.285005     0.0000000</span></span>
<span><span class="co"># Surveillance4    8.723064      4.462216      3.529919     0.7661988</span></span>
<span><span class="co"># Surveillance5    5.735601      7.660914      8.219343     3.6764471</span></span>
<span><span class="co"># Activated1      13.835812     10.549585      7.522241     3.5081678</span></span>
<span><span class="co"># Activated2      22.369534     18.084185     13.030198     8.7718675</span></span>
<span><span class="co"># Activated3      17.792197     20.061153     17.297088    11.6027919</span></span>
<span><span class="co"># Phagocytosis    26.392094     29.687218     25.688176    19.5139817</span></span>
<span><span class="co">#               Surveillance4 Surveillance5 Activated1 Activated2 Activated3</span></span>
<span><span class="co"># Homeostasis       8.7230645      5.735601  13.835812  22.369534  17.792197</span></span>
<span><span class="co"># Surveillance1     4.4622158      7.660914  10.549585  18.084185  20.061153</span></span>
<span><span class="co"># Surveillance2     3.5299193      8.219343   7.522241  13.030198  17.297088</span></span>
<span><span class="co"># Surveillance3     0.7661988      3.676447   3.508168   8.771868  11.602792</span></span>
<span><span class="co"># Surveillance4     0.0000000      1.301993   1.908912   6.800320   8.081606</span></span>
<span><span class="co"># Surveillance5     1.3019930      0.000000   2.087510   6.593177   5.495311</span></span>
<span><span class="co"># Activated1        1.9089124      2.087510   0.000000   2.014579   3.216930</span></span>
<span><span class="co"># Activated2        6.8003204      6.593177   2.014579   0.000000   1.950235</span></span>
<span><span class="co"># Activated3        8.0816058      5.495311   3.216930   1.950235   0.000000</span></span>
<span><span class="co"># Phagocytosis     15.4674423     12.337110   8.521408   5.171625   2.154714</span></span>
<span><span class="co">#               Phagocytosis</span></span>
<span><span class="co"># Homeostasis      26.392094</span></span>
<span><span class="co"># Surveillance1    29.687218</span></span>
<span><span class="co"># Surveillance2    25.688176</span></span>
<span><span class="co"># Surveillance3    19.513982</span></span>
<span><span class="co"># Surveillance4    15.467442</span></span>
<span><span class="co"># Surveillance5    12.337110</span></span>
<span><span class="co"># Activated1        8.521408</span></span>
<span><span class="co"># Activated2        5.171625</span></span>
<span><span class="co"># Activated3        2.154714</span></span>
<span><span class="co"># Phagocytosis      0.000000</span></span></code></pre>
</details><p>Next, we perform a similar analysis on the perturbed Seurat assay
‘M4_down’</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set the default assay to the perturbed assay and prepare the data</span></span>
<span><span class="va">seurat_obj_perturbed</span> <span class="op">&lt;-</span> <span class="va">seurat_obj</span> <span class="co"># set up a new seurat object to avoid confusion</span></span>
<span></span>
<span><span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">seurat_obj_perturbed</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"M4_down"</span></span>
<span></span>
<span><span class="va">seurat_obj_perturbed</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span><span class="va">seurat_obj_perturbed</span>, selection.method <span class="op">=</span> <span class="st">"vst"</span>, nfeatures <span class="op">=</span> <span class="fl">2000</span><span class="op">)</span></span>
<span><span class="va">seurat_obj_perturbed</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span><span class="va">seurat_obj_perturbed</span><span class="op">)</span></span></code></pre></div>
<p>Run PCA for the perturbed data and please remember to change the
‘reduction.name’</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seurat_obj_perturbed</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span><span class="va">seurat_obj_perturbed</span>, features <span class="op">=</span> <span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/VariableFeatures.html" class="external-link">VariableFeatures</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">seurat_obj_perturbed</span><span class="op">)</span>, reduction.name <span class="op">=</span> <span class="st">"M4_down_pca"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">seurat_obj_perturbed</span><span class="op">@</span><span class="va">reductions</span></span></code></pre></div>
<details><summary>
Output
</summary><pre><code>&gt; seurat_obj_perturbed@reductions
$umap
A dimensional reduction object with key UMAP_
 Number of dimensions: 2
 Projected dimensional reduction calculated:  FALSE
 Jackstraw run: FALSE
 Computed using assay: RNA

$pca
A dimensional reduction object with key PC_
 Number of dimensions: 50
 Projected dimensional reduction calculated:  FALSE
 Jackstraw run: FALSE
 Computed using assay: RNA

$harmony
A dimensional reduction object with key harmony_
 Number of dimensions: 50
 Projected dimensional reduction calculated:  TRUE
 Jackstraw run: FALSE
 Computed using assay: RNA

$M4_down_pca
A dimensional reduction object with key m4_down_pca_
 Number of dimensions: 50
 Projected dimensional reduction calculated:  FALSE
 Jackstraw run: FALSE
 Computed using assay: M4_down
</code></pre>
</details><p>Compute similarity distance for the perturbed data</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># ComputeDistance</span></span>
<span><span class="va">df_edist_perturbed</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ComputeDistance.html">ComputeDistance</a></span><span class="op">(</span><span class="va">seurat_obj_perturbed</span>, groupby <span class="op">=</span> <span class="st">'functional_anno'</span>, reduction <span class="op">=</span> <span class="st">'M4_down_pca'</span>, method <span class="op">=</span> <span class="st">"edist"</span>, verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">df_edist_perturbed</span></span>
<span></span>
<span></span>
<span><span class="co"># Rearrange rows and columns of df_edist_observed based on the custom order</span></span>
<span><span class="va">df_edist_perturbed_reordered</span> <span class="op">&lt;-</span> <span class="va">df_edist_perturbed</span><span class="op">[</span><span class="va">custom_order</span>, <span class="va">custom_order</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># View the reordered dataframe</span></span>
<span><span class="va">df_edist_perturbed_reordered</span></span></code></pre></div>
<details><summary>
Output
</summary><pre><code>&gt; df_edist_perturbed
              Phagocytosis Surveillance3 Surveillance2 Activated1 Surveillance1
Phagocytosis     0.0000000     4.1887166     5.3468811  1.9114647      6.367927
Surveillance3    4.1887166     0.0000000     0.3544971  0.8140641      0.702068
Surveillance2    5.3468811     0.3544971     0.0000000  1.6971191      0.444851
Activated1       1.9114647     0.8140641     1.6971191  0.0000000      2.354567
Surveillance1    6.3679270     0.7020680     0.4448510  2.3545673      0.000000
Surveillance4    3.3193247     0.2185912     0.8524824  0.3940233      1.108151
Activated2       1.6043523     1.2675249     2.0982363  0.3861708      2.864140
Surveillance5    2.9806149     0.5994561     1.5770397  0.4120399      1.629711
Homeostasis      5.0451679     1.9123500     2.8172922  2.0402917      1.892958
Activated3       0.9335111     1.9314490     3.0544205  0.5010339      3.670300
              Surveillance4 Activated2 Surveillance5 Homeostasis Activated3
Phagocytosis      3.3193247  1.6043523     2.9806149    5.045168  0.9335111
Surveillance3     0.2185912  1.2675249     0.5994561    1.912350  1.9314490
Surveillance2     0.8524824  2.0982363     1.5770397    2.817292  3.0544205
Activated1        0.3940233  0.3861708     0.4120399    2.040292  0.5010339
Surveillance1     1.1081507  2.8641397     1.6297111    1.892958  3.6702999
Surveillance4     0.0000000  0.9458629     0.2065312    1.271465  1.2788325
Activated2        0.9458629  0.0000000     0.8679631    2.958652  0.3551624
Surveillance5     0.2065312  0.8679631     0.0000000    1.029319  0.9134280
Homeostasis       1.2714649  2.9586525     1.0293188    0.000000  2.7488030
Activated3        1.2788325  0.3551624     0.9134280    2.748803  0.0000000
</code></pre>
</details><p>We visualize the similarity distance matrices for both the original
and perturbed data using a heatmap. This will allow us to compare the
functional clusters’ similarity or difference before and after the
perturbation</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#---------------------------</span></span>
<span><span class="co"># NOTE plot heatmap</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/hadley/reshape" class="external-link">reshape2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://scales.r-lib.org" class="external-link">scales</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span>  <span class="co"># For arranging plots side by side</span></span>
<span></span>
<span></span>
<span><span class="co"># # Example usage with default color palette</span></span>
<span><span class="va">order_level</span> <span class="op">&lt;-</span> <span class="va">custom_order</span></span>
<span></span>
<span><span class="co"># HeatmapDistance()</span></span>
<span><span class="va">ht</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/HeatmapDistance.html">HeatmapDistance</a></span><span class="op">(</span>df_original <span class="op">=</span> <span class="va">df_edist_original</span>, df_perturbed <span class="op">=</span> <span class="va">df_edist_perturbed</span>, custom_order <span class="op">=</span> <span class="va">order_level</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ht</span></span></code></pre></div>
<p><img src="figures/basic_tutorial/edistance_after_M4_down.png" width="5000" height="2500"></p>
<p>We can see that the distance is smaller in the heatmap on the right,
which indicates that the activated cells have become transcriptomically
more similar to the homeostatic cells after our in-silco
perturbation.</p>
</div>
<div class="section level3">
<h3 id="apply-a-knock-in">Apply a knock-in<a class="anchor" aria-label="anchor" href="#apply-a-knock-in"></a>
</h3>
<p>In this step, we will simulate another in-silico perturbation with
<code>ModulePerturbation</code> to up-regulate the same co-expression
module, and we will visualize the resulting transition vector field with
<code>PlotTransitionVectors</code>. The only difference here is that we
set <code>perturb_dir=1</code> for an up-regulation whereas previously
we had <code>perturb_dir=-1</code> for a down-regulation.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># apply the up-regulation of module M4</span></span>
<span><span class="va">perturbation_name</span> <span class="op">&lt;-</span> <span class="st">'M4_up'</span></span>
<span><span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ModulePerturbation.html">ModulePerturbation</a></span><span class="op">(</span></span>
<span>    <span class="va">seurat_obj</span>,</span>
<span>    mod <span class="op">=</span> <span class="st">"MG-M4"</span>,</span>
<span>    perturb_dir <span class="op">=</span> <span class="fl">1</span>, <span class="co"># positive value indicates knock-in, all </span></span>
<span>    perturbation_name <span class="op">=</span> <span class="va">perturbation_name</span>,</span>
<span>    graph <span class="op">=</span> <span class="st">'RNA_nn'</span>, </span>
<span>    n_hubs <span class="op">=</span> <span class="fl">10</span>, </span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot the transition vectors</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PlotTransitionVectors.html">PlotTransitionVectors</a></span><span class="op">(</span></span>
<span>    <span class="va">seurat_obj</span>,</span>
<span>    perturbation_name <span class="op">=</span> <span class="st">"M4_up"</span>,</span>
<span>    reduction <span class="op">=</span> <span class="st">'umap'</span>,</span>
<span>    color.by<span class="op">=</span><span class="st">'pseudotime'</span>,</span>
<span>    grid_resolution <span class="op">=</span> <span class="fl">25</span>,</span>
<span>    point_size<span class="op">=</span><span class="fl">2</span> <span class="op">)</span> <span class="op">+</span> </span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoLegend</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_color_gradientn</a></span><span class="op">(</span>colors<span class="op">=</span><span class="fu">viridis</span><span class="fu">::</span><span class="fu"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">plasma</a></span><span class="op">(</span><span class="fl">256</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_equal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"M4 knock-in"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot the transition vectors</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PlotTransitionVectors.html">PlotTransitionVectors</a></span><span class="op">(</span></span>
<span>    <span class="va">seurat_obj</span>,</span>
<span>    perturbation_name <span class="op">=</span> <span class="st">"M4_down"</span>,</span>
<span>    reduction <span class="op">=</span> <span class="st">'umap'</span>,</span>
<span>    color.by<span class="op">=</span><span class="st">'pseudotime'</span>,</span>
<span>    grid_resolution <span class="op">=</span> <span class="fl">25</span>,</span>
<span>    point_size<span class="op">=</span><span class="fl">2</span> <span class="op">)</span> <span class="op">+</span> </span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoLegend</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_color_gradientn</a></span><span class="op">(</span>colors<span class="op">=</span><span class="fu">viridis</span><span class="fu">::</span><span class="fu"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">plasma</a></span><span class="op">(</span><span class="fl">256</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_equal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"M4 knock-down"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p1</span> <span class="op">|</span> <span class="va">p2</span></span></code></pre></div>
<p><img src="figures/basic_tutorial/vectorfield_compare_MG-M4_velocyto.png" width="1000" height="1000"></p>
<p>As expected, we can broadly see that the transition vectors indicate
opposite affects of the knock-in and knock-down on the predicted cell
states.</p>
</div>
</div>
<div class="section level2">
<h2 id="conclusion">Conclusion<a class="anchor" aria-label="anchor" href="#conclusion"></a>
</h2>
<p>Here we demonstrated the basics of the
<strong><code>compact</code></strong> method to apply an in-silico
knockdown of a the microglial activation module M4. By down-regulating
this module, the method predicted that cells would shift to a more
homeostatic-like state.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://smorabit.github.io" class="external-link">Sam Morabito</a>, <a href="https://github.com/rootze" class="external-link">Zechuan Shi</a>, <a href="https://swaruplab.bio.uci.edu/" class="external-link">Swarup Lab</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
