<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>transcription factor perturbation analysis • compact</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="transcription factor perturbation analysis">
<meta name="description" content="Tutorial for performing transcription factor perturbation analysis with compact.
">
<meta property="og:description" content="Tutorial for performing transcription factor perturbation analysis with compact.
">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">compact</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/TF_tutorial.html">transcription factor perturbation analysis</a></li>
    <li><a class="dropdown-item" href="../articles/basic_tutorial.html">co-expression module perturbation analysis</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>transcription factor perturbation analysis</h1>
            
      

      <div class="d-none name"><code>TF_tutorial.Rmd</code></div>
    </div>

    
    
<p>Compiled: 28-10-2024</p>
<p>Source: <code>vignettes/basic_tutorial.Rmd</code></p>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This tutorial covers the basics of using
<strong><code>compact</code></strong> to perform transcription factor
(TF) perturbation analysis in single-cell transcriptomic data.</p>
<p>For this tutorial we will use an integrated single-nucleus RNA-seq
dataset of microglia from three different studies of Alzheimer’s
disease, as we described previously in the <a href="https://www.sciencedirect.com/science/article/pii/S2667237523001273?via%3Dihub" class="external-link">hdWGCNA
paper</a>.</p>
<p>The in-silico perturbations that we demonstrate in this tutorial are
based on a TF regulatory networks, which represent regulatory
relationships between TFs and downstream target genes. In order to run
these pertubations, we first need to perform co-expression network
analysis and group genes into modules by running <a href="https://smorabit.github.io/hdWGCNA/articles/basic_tutorial.html" class="external-link">hdWGCNA</a>,
and then we need to <a href="https://smorabit.github.io/hdWGCNA/articles/tf_network.html" class="external-link">construct
a TF regulatory network</a>. For this tutorial, co-expression network
analysis has already been run but we will demonstrate constructing the
TF regulatory network.</p>
</div>
<div class="section level2">
<h2 id="load-the-dataset">Load the dataset<a class="anchor" aria-label="anchor" href="#load-the-dataset"></a>
</h2>
<p>Download the tutorial dataset from these Google Drive links:</p>
<ul>
<li><p><a href="https://drive.google.com/file/d/15-0URxGUgE2eCxx15zkoVwkkmLhZTFKH/view?usp=drive_link" class="external-link">Seurat
object</a></p></li>
<li><p><a href="https://drive.google.com/file/d/1BcnUx3eh4zHwk_JfBo8zpB70ewp0u1tX/view?usp=drive_link" class="external-link">co-expression
adjacency matrix (.rda)</a></p></li>
</ul>
<p>First we load the required R libraries and the tutorial dataset. When
loading the tutorial dataset, you need to update the filepath to the
co-expression adjacency matrix (TOM) within the Seurat object (shown
below).</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/" class="external-link">cowplot</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">velocyto.R</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://smorabit.github.io/hdWGCNA/" class="external-link">hdWGCNA</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://smorabit.github.io/COMPACT/" class="external-link">compact</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/get_theme.html" class="external-link">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/theme_cowplot.html" class="external-link">theme_cowplot</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># set random seed for reproducibility</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">12345</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># load the processed alzheimer's microglia dataset</span></span>
<span><span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span>file<span class="op">=</span><span class="st">'AD_MG_hdWGCNA_COMPACT_tutorial.rds'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># update the path to the co-expression network adjacency matrix</span></span>
<span><span class="va">net</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://smorabit.github.io/hdWGCNA/reference/GetNetworkData.html" class="external-link">GetNetworkData</a></span><span class="op">(</span><span class="va">seurat_obj</span><span class="op">)</span></span>
<span><span class="va">net</span><span class="op">$</span><span class="va">TOMFiles</span> <span class="op">&lt;-</span> <span class="st">'AD_MG_TOM.rda'</span> <span class="co"># it is better to use the absolute path</span></span>
<span><span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://smorabit.github.io/hdWGCNA/reference/SetNetworkData.html" class="external-link">SetNetworkData</a></span><span class="op">(</span><span class="va">seurat_obj</span>, <span class="va">net</span><span class="op">)</span></span></code></pre></div>
<p>If you have not already done so, we need to construct a
nearest-neighbors graph in the Seurat object. This is a key step in the
typical single-cell clustering pipeline.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a>seurat_obj <span class="ot">&lt;-</span> <span class="fu">FindNeighbors</span>(</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a>    seurat_obj,</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a>    <span class="at">reduction=</span><span class="st">'harmony'</span>,nearest neighbors?</span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a>    <span class="at">assay =</span> <span class="st">'RNA'</span>,</span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a>    <span class="at">annoy.metric =</span> <span class="st">'cosine'</span></span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="construct-tf-regulatory-network">Construct TF Regulatory Network<a class="anchor" aria-label="anchor" href="#construct-tf-regulatory-network"></a>
</h2>
<p>First we construct a TF regulatory network using hdWGCNA by following
<a href="https://smorabit.github.io/hdWGCNA/articles/tf_network.html#transcription-factor-network-analysis" class="external-link">this
tutorial</a>, and we first described this approach in our paper <a href="https://www.cell.com/cell-reports/fulltext/S2211-1247(24)00284-5" class="external-link">Childs
&amp; Morabito et al., Cell Reports (2024)</a>. We include the code
below, but please visit the <a href="https://smorabit.github.io/hdWGCNA/articles/tf_network.html#transcription-factor-network-analysis" class="external-link">hdWGCNA
TF network tutorial</a> for further explanations about this
analysis.</p>
<details><summary>
Run the TF regulatory network code
</summary><p>Part 1: Identify TFs in promoter regions</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load these packages into R:</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://jaspar.genereg.net/" class="external-link">JASPAR2020</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">motifmatchr</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ge11232002/TFBSTools" class="external-link">TFBSTools</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">EnsDb.Hsapiens.v86</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">BSgenome.Hsapiens.UCSC.hg38</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicRanges" class="external-link">GenomicRanges</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/dmlc/xgboost" class="external-link">xgboost</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># get the set of Motifs from JASPAR2020</span></span>
<span><span class="va">pfm_core</span> <span class="op">&lt;-</span> <span class="fu">TFBSTools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/TFBSTools/man/getMatrixSet-methods.html" class="external-link">getMatrixSet</a></span><span class="op">(</span></span>
<span>  x <span class="op">=</span> <span class="va">JASPAR2020</span>,</span>
<span>  opts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>collection <span class="op">=</span> <span class="st">"CORE"</span>, tax_group <span class="op">=</span> <span class="st">'vertebrates'</span>, all_versions <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># scan promoter regions for TF motifs</span></span>
<span><span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://smorabit.github.io/hdWGCNA/reference/MotifScan.html" class="external-link">MotifScan</a></span><span class="op">(</span></span>
<span>  <span class="va">seurat_obj</span>,</span>
<span>  species_genome <span class="op">=</span> <span class="st">'hg38'</span>,</span>
<span>  pfm <span class="op">=</span> <span class="va">pfm_core</span>,</span>
<span>  EnsDb <span class="op">=</span> <span class="va">EnsDb.Hsapiens.v86</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Part 2: Construct TF Regulatory Network</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get the motif df:</span></span>
<span><span class="va">motif_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://smorabit.github.io/hdWGCNA/reference/GetMotifs.html" class="external-link">GetMotifs</a></span><span class="op">(</span><span class="va">seurat_obj</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># keep all TFs, and then remove all genes from the grey module</span></span>
<span><span class="va">tf_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">motif_df</span><span class="op">$</span><span class="va">gene_name</span><span class="op">)</span></span>
<span><span class="va">modules</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://smorabit.github.io/hdWGCNA/reference/GetModules.html" class="external-link">GetModules</a></span><span class="op">(</span><span class="va">seurat_obj</span><span class="op">)</span></span>
<span><span class="va">nongrey_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">modules</span>, <span class="va">module</span> <span class="op">!=</span> <span class="st">'grey'</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="va">.</span><span class="op">$</span><span class="va">gene_name</span></span>
<span><span class="va">genes_use</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">tf_genes</span>, <span class="va">nongrey_genes</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># update the gene list and re-run SetDatExpr</span></span>
<span><span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://smorabit.github.io/hdWGCNA/reference/SetWGCNAGenes.html" class="external-link">SetWGCNAGenes</a></span><span class="op">(</span><span class="va">seurat_obj</span>, <span class="va">genes_use</span><span class="op">)</span></span>
<span><span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://smorabit.github.io/hdWGCNA/reference/SetDatExpr.html" class="external-link">SetDatExpr</a></span><span class="op">(</span><span class="va">seurat_obj</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># define model params:</span></span>
<span><span class="va">model_params</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    objective <span class="op">=</span> <span class="st">'reg:squarederror'</span>,</span>
<span>    max_depth <span class="op">=</span> <span class="fl">1</span>,</span>
<span>    eta <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>    nthread<span class="op">=</span><span class="fl">8</span>,</span>
<span>    alpha<span class="op">=</span><span class="fl">0.5</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># construct the TF network</span></span>
<span><span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://smorabit.github.io/hdWGCNA/reference/ConstructTFNetwork.html" class="external-link">ConstructTFNetwork</a></span><span class="op">(</span><span class="va">seurat_obj</span>, <span class="va">model_params</span><span class="op">)</span></span></code></pre></div>
<p>Part 3: Define TF Regulons</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># define the TF regulons</span></span>
<span><span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://smorabit.github.io/hdWGCNA/reference/AssignTFRegulons.html" class="external-link">AssignTFRegulons</a></span><span class="op">(</span></span>
<span>    <span class="va">seurat_obj</span>,</span>
<span>    strategy <span class="op">=</span> <span class="st">"A"</span>,</span>
<span>    reg_thresh <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>    n_tfs <span class="op">=</span> <span class="fl">10</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># calculate expression scores for TF regulonspositive regulons</span></span>
<span><span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://smorabit.github.io/hdWGCNA/reference/RegulonScores.html" class="external-link">RegulonScores</a></span><span class="op">(</span></span>
<span>    <span class="va">seurat_obj</span>,</span>
<span>    target_type <span class="op">=</span> <span class="st">'positive'</span>,</span>
<span>    ncores<span class="op">=</span><span class="fl">8</span></span>
<span><span class="op">)</span></span>
<span><span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://smorabit.github.io/hdWGCNA/reference/RegulonScores.html" class="external-link">RegulonScores</a></span><span class="op">(</span></span>
<span>    <span class="va">seurat_obj</span>,</span>
<span>    target_type <span class="op">=</span> <span class="st">'negative'</span>,</span>
<span>    cor_thresh <span class="op">=</span> <span class="op">-</span><span class="fl">0.05</span>,</span>
<span>    ncores<span class="op">=</span><span class="fl">8</span></span>
<span><span class="op">)</span></span></code></pre></div>
</details>
</div>
<div class="section level2">
<h2 id="perform-in-silico-tf-perturbations">Perform in-silico TF perturbations<a class="anchor" aria-label="anchor" href="#perform-in-silico-tf-perturbations"></a>
</h2>
<p>Resulting from the previous step, we now have the TF regulatory
network for the microglia dataset. Let’s briefly check some of these
results before performing perturbation analysis.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://smorabit.github.io" class="external-link">Sam Morabito</a>, <a href="https://github.com/rootze" class="external-link">Zechuan Shi</a>, <a href="https://swaruplab.bio.uci.edu/" class="external-link">Swarup Lab</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
